cmake/compiler_config.cmake

```diff
@@ -0,0 +1,79 @@
+#------------------------------------------------------------------------------
+# compiler flags/definitions
+#------------------------------------------------------------------------------
+set(gnu_clang_flags
+    # we want to be as strict as possible
+    -Wall
+    $<$<BOOL:${BUILD_STRICT_MODE}>:-Werror>
+    # optimization
+    -msse3
+    # disable warnings
+    -Wno-deprecated
+    -Wno-deprecated-declarations
+    -Wno-unused-local-typedefs
+)
+
+set(msvc_flags
+    # we want to be as strict as possible
+    /W3
+    $<$<BOOL:${BUILD_STRICT_MODE}>:/WX>
+    # enable pdb generation.
+    /Zi
+    # standards compliant.
+    /Zc:inline
+    /Zc:rvalueCast
+    # enable multiprocessor builds.
+    /MP
+    # disable warnings
+    /wd4244
+    /wd4267
+    /wd4273
+    /wd4305
+    /wd4506
+    /wd4996
+    /wd4180
+)
+
+set(msvc_definitions
+    # Make sure WinDef.h does not define min and max macros which
+    # will conflict with std::min() and std::max().
+    NOMINMAX
+
+    _CRT_SECURE_NO_WARNINGS
+    _SCL_SECURE_NO_WARNINGS
+
+    # Boost
+    BOOST_ALL_DYN_LINK
+    BOOST_CONFIG_SUPPRESS_OUTDATED_MESSAGE
+)
+
+#------------------------------------------------------------------------------
+# compiler configuration
+#------------------------------------------------------------------------------
+# Don not use GNU extension 
+# Use -std=c++11 instead of -std=gnu++11
+set(CMAKE_CXX_EXTENSIONS OFF)
+
+function(mayaUsd_compile_config TARGET)
+    # required compiler feature
+    target_compile_features(${TARGET} 
+        PRIVATE
+            $<$<VERSION_LESS:MAYA_APP_VERSION,2019>:cxx_std_11>
+            $<$<VERSION_GREATER_EQUAL:MAYA_APP_VERSION,2019>:cxx_std_14>
```

Comment:
I'm a little conflicted on burying the C++ language standard inside `mayaUsd_compile_config()`. It feels like this should be more of a repo-level setting. Do we envision having different parts of the repo on different standards? That would seem a bit weird to me.

And somewhat related to that, which parts of the tree do we expect would *not* call `mayaUsd_compile_config()` for their targets? For now, maybe that's the AL and pxr plugins, but do we expect that to be the case long term? I'm hoping to make some time in the not too distant future to remove all the pxr build stuff from the remaining pxr plugin code.

Reply:
> I'm a little conflicted on burying the C++ language standard inside mayaUsd_compile_config(). 

There are two ways of adding C++ standard in modern CMake.

1- Setting the C++ standard directly before any targets being defined which is the approach you have in your PR:

```
set(CMAKE_CXX_STANDARD 11)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```

This setting applies to all targets `globally` which in this case is Ok because we obviously want all targets to use the exact same standard. 

2- Setting the C++ standard based on features which is a more finer grained approach

target_compile_features(foo PRIVATE|PUBLIC|INTERFACE cxx_std_11)

Now, there reason I went with this approach is because  `mayaUsd_compile_config` already use `target_compile_options` and `target_compile_definitions` and `target_compile_features` seems like a good fit there.

To be clear, every targets **must** add `mayaUsd_compile_config`.  Notice that all the entries for target_compile_options, target_compile_definitions, target_compile_features are declared as **PRIVATE** which is very important.  Any change in `mayaUsd_compile_config` affects **all** targets that are inheriting this function which simulates a global behavior. 

>  It feels like this should be more of a repo-level setting. Do we envision having different parts of the repo on different standards? That would seem a bit weird to me.

Not at all, all targets should have the same standard. 

> And somewhat related to that, which parts of the tree do we expect would not call mayaUsd_compile_config() for their targets?

Every targets **must** add `mayaUsd_compile_config`

Reply:
Right. I guess I was reacting to two things:
- The fact that we're already breaking the "every target must add `mayaUsd_compile_config()`" rule, since the AL and pxr plugin code do not use it. But that's probably a whole other can of beans, and they probably should use it eventually.
- But more generally, we're saying that **every** target **must** add `mayaUsd_compile_config()` so that it picks up (effectively) global settings. It seems a little counter-intuitive to me to require authors to do something in their CMakeLists.txt so that they pick up the settings all targets need?

Anyway, it doesn't bother me that much, and I'm likely missing other context here. Just wanted to raise the question. Thanks for humoring me!